### PR TITLE
phase2: use the v1alpha2 kubeadm config if needed 

### PR DESCRIPTION
**Pull request by neolit123**
Refs kubernetes/kubernetes#66338

kubeadm 1.11 needs v1alpha2, thus we should use `config migrate`
to update the configration file. this should also work for future
versions, as long as `config migrate` works.

kubeadm 1.10.x would not trow an error.

https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-gce-selfhosting/537/artifacts/e2e-537-f2b95-master/serial-1.log

```
Jul 19 19:33:35 e2e-537-f2b95-master startup-script: INFO startup-script: your configuration file uses an old API spec: "kubeadm.k8s.io/v1alpha1". Please use kubeadm v1.11 instead and run 'kubeadm config migrate --old-config old.yaml --new-config new.yaml', which will write the new, similar spec using a newer API version.
Jul 19 19:33:35 e2e-537-f2b95-master startup-script: INFO startup-script: Return code 1.
```

@timothysc @dims @BenTheElder

### CODE REVIEW

File: phase2/kubeadm/configure-vm-kubeadm-master.sh [hunk 1]
```diff
@@ -17,14 +17,20 @@ fi
 mkdir -p $KUBEADM_DIR
 cat <<EOF |tee $KUBEADM_CONFIG_FILE
 kind: MasterConfiguration
-apiVersion: kubeadm.k8s.io/v1alpha1
+apiVersion: kubeadm.k8s.io/v1alpha2
```
Comment (BenTheElder) [inline phase2/kubeadm/configure-vm-kubeadm-master.sh]: hmm, I guess we will need to use an old kubernetes-anywhere on the older kubernetes release branches..
Can we make this conditional on the kubeadm or kubernetes veresion?

**Comment by neolit123** [comment]
updated, after the suggestion to use `config migrate` by @dims

**Comment by neolit123** [comment]
@luxas that was the first version of this patch, but then both @dims and @BenTheElder said that kubernetes-everywhere can be used by people with k8s <1.11 versions and v1alpha1 support still has to be there.

thoughts on that?

**Comment by luxas** [comment]
(Preferably also branched per version as @bentheelder said)

**Comment by luxas** [comment]
Because kubeadm at HEAD can't upgrade v1alpha1 configs at all.

**Comment by neolit123** [comment]
ok, i can make the change.

BTW, i tested it last night, for the sake of it with HEAD, and `migrate` did work on this simple `v1alpha1` that's inlined in the `.sh.` script. that's because of the v1alpha1 leftovers, i assume.

File: phase2/kubeadm/configure-vm-kubeadm-master.sh [hunk 2]
```diff
@@ -59,4 +59,7 @@ featureGates:
 EOF
 fi
 
+# first migrate the configuration from `v1alpha1` to any newer version.
+# don't error if `config migrate` is not available as a command.
+kubeadm config migrate --old-config $KUBEADM_CONFIG_FILE --new-config $KUBEADM_CONFIG_FILE || true
```
Comment (timothysc) [inline phase2/kubeadm/configure-vm-kubeadm-master.sh]: I think we should just upload the new config v1alpha2, and not call migrate imo.

File: phase2/kubeadm/configure-vm-kubeadm-master.sh [hunk 3]
```diff
@@ -15,24 +15,67 @@ elif [[ "$KUBEADM_CNI_PLUGIN" == "weave" ]]; then
...
 networking:
   podSubnet: "$POD_NETWORK_CIDR"
-kubernetesVersion: "$KUBEADM_KUBERNETES_VERSION"
-token: "$KUBEADM_TOKEN"
 EOF
 
-if [[ "$KUBEADM_ENABLE_CLOUD_PROVIDER" == true ]]; then
+# handle v1alpha1
+########################################################################
+if [ "$KUBEADM_KUBERNETES_VERSION_MINOR" -lt "11" ]; then
```
Comment (BenTheElder) [inline phase2/kubeadm/configure-vm-kubeadm-master.sh]: nit: the rest of this script looks to be bash leveraging `[[`, why not here for consistency?

**Comment by BenTheElder** [comment]
Quick check on my way out says it looks good to me, one nit. We'll need to update the kubeadm image in test-infra to consume this ref once it's merged.

**Comment by neolit123** [comment]
@BenTheElder patch updated.
> We'll need to update the kubeadm image in test-infra to consume this ref once it's merged.

let me know if i can help there, preferably a small guide would help me - e.g. which file, and what has to be changed. thanks.

**Comment by BenTheElder** [comment]
@neolit123 thanks! the image obtains kubernetes-anywhere here https://github.com/kubernetes/test-infra/blob/0e4d5ea4ff95460f5aa08baea6d17af377d7a68a/images/kubeadm/runner#L29 

This might not be the best way, but that's how it works currently. I'm not sure who maintains it these days..

**Comment by neolit123** [comment]
@BenTheElder 
got it! will update the SHA, once the current PR is approved.

File: phase2/kubeadm/configure-vm-kubeadm-master.sh [hunk 4]
```diff
@@ -15,24 +15,67 @@ elif [[ "$KUBEADM_CNI_PLUGIN" == "weave" ]]; then
...
+if [[ "$KUBEADM_KUBERNETES_VERSION_MINOR" -lt "11" ]]; then
+
+  # add api version and token
   cat <<EOF |tee -a $KUBEADM_CONFIG_FILE
+apiVersion: kubeadm.k8s.io/v1alpha1
+token: "$KUBEADM_TOKEN"
+EOF
+
+  # add cloud provider
+  if [[ "$KUBEADM_ENABLE_CLOUD_PROVIDER" == true ]]; then
```
Comment (luxas) [inline phase2/kubeadm/configure-vm-kubeadm-master.sh]: nit: move this out of the two if blocks wrt v1alpha1 and v1alpha2

**Comment by neolit123** [comment]
patch updated. moved to using `--cloud-provider` from `apiServerExtraArgs` for both config variants.

**Comment by neolit123** [comment]
updated.

**Pull request merged**